python/rascaline/utils/power_spectrum.py

```diff
@@ -0,0 +1,320 @@
+                    values=np.array(
+                        [
+                            properties_1.tolist() + properties_2.tolist()
+                            for properties_1 in block_1.properties.values
+                            for properties_2 in block_2.properties.values
+                        ],
+                        dtype=np.int32,
+                    ),
+                )
+
+                # Compute the invariants by summation and store the results this is
+                # equivalent to an einsum with: ima, imb -> iab
+                data = factor * np.matmul(block_1.values.swapaxes(1, 2), block_2.values)
+
+                new_block = TensorBlock(
+                    values=data.reshape(data.shape[0], -1),
+                    samples=block_1.samples,
+                    components=[],
+                    properties=properties,
+                )
+
+                for parameter in block_1.gradients_list():
+                    if parameter == "positions":
+                        _positions_gradients(new_block, block_1, block_2, factor)
+
+                keys.append((ell, species_center))
+                blocks.append(new_block)
+
+        keys = Labels(
+            names=["l", "species_center"],
+            values=np.array(keys, dtype=np.int32),
+        )
+
+        return TensorMap(keys, blocks).keys_to_properties("l")
+
+
+def _positions_gradients(new_block, block_1, block_2, factor):
+    gradient_1 = block_1.gradient("positions")
+    gradient_2 = block_2.gradient("positions")
+
+    if len(gradient_1.samples) == 0 or len(gradient_2.samples) == 0:
+        gradients_samples = Labels.empty()
+        gradient_values = np.array([]).reshape(0, 1, len(new_block.properties))
+    else:
+        # The "sample" dimension in the power spectrum gradient samples do
+        # not necessarily matches the "sample" dimension in the spherical
+        # expansion gradient samples. We create new samples by creating a
+        # union between the two gradient samples.
+        gradients_samples_values = np.unique(
+            np.concatenate([gradient_1.samples, gradient_2.samples])
+        )
+
+        gradients_samples = Labels(
+            names=gradient_1.samples.names,
+            values=np.asarray(gradients_samples_values.tolist()),
+        )
+
+        gradient_values = np.zeros(
+            [gradients_samples.shape[0], 3, len(new_block.properties)]
+        )
+
+        # this is equivalent to an einsum with: ixma, imb -> ixab
+        gradient_1_values = factor * np.matmul(
+            gradient_1.values.swapaxes(2, 3),
+            block_2.values[gradient_1.samples["sample"], np.newaxis, :, :],
+        )
+
+        # Find correct indices of the 1st gradient values and fill the
+        # values in
+        _, grad1_sample_idxs, _ = np.intersect1d(
+            gradients_samples, gradient_1.samples, return_indices=True
+        )
```

Comment:
I'm surprised this works with the latest version of equistore from #194? `gradients.samples` is no longer a np.ndarray